### ISSUE TITLE
Audio channel conversion optimization

### ISSUE THREAD

**Issue by JoelLinn** [libsdl-org/SDL#6560]
As part of the audio channel conversion rewrite, some intrinsics optimizations have been removed while some SSE variants were kept https://github.com/libsdl-org/SDL/blob/fe396e306effb170b1abd093111d3f5c5338ae4b/src/audio/SDL_audiocvt.c#L754-L767 
What was the thinking behind which to remove/keep? They should be low maintenance once written and tested.
In case it is generally not desired to blow SDL with intrinsics code, it could be possible to write a generator similar to gen_audio_channel_conversion.c that also emits vectorized code using the [Halide](https://halide-lang.org/) DSL

**Comment by flibitijibibo** [comment]
I believe the rationale was that the new matrix coefficients caused some of the math to no longer line up with the SIMD implementations; some of the SSE paths remained okay but I think the AVX work got cut:

https://github.com/libsdl-org/SDL/pull/5934#issuecomment-1190577520

Sadly I don't have anything to replace these in FAudio's SIMD work, even the generic mix functions assume some things SDL doesn't care about:

https://github.com/FNA-XNA/FAudio/blob/master/src/FAudio_internal_simd.c

**Comment by flibitijibibo** [comment]
Took another look since it's been a while, it was just 5.1->stereo that got cut. This is the new scalar formula:

https://github.com/libsdl-org/SDL/blob/main/src/audio/SDL_audio_channel_converters.h#L923

**Comment by icculus** [comment]
I spent hours trying to write the AVX 5.1-to-stereo code. It doesn't work when moving between these two formats...you end up doing an enormous amount of permutes to shuffle data to the right positions and you lose all your gains. The simple scalar for-loop turned out to be faster.

The old version was doing it incorrectly, so it was very fast and completely incorrect output.

I was so mad about my failure to make this work, I even wrote a blog post about it, lol:

https://cohost.org/icculus/post/59735-assumption-based-opt

**Comment by JoelLinn** [comment]
Thanks for elaborating and putting the effort into it. I kind of feel sorry for contributing the AVX code which coerced you into trying. 

I looked into 5.1->2 again and played around with some  Halide schedules but of cause LLVM also created a permute hell.
Interestingly though I was able to make it about 2 times faster using only SSE 4.1. For separate input and output buffers (not what's needed in SDL) the SDL scalar code also was about 1.5 times faster - much closer to the permute hell. Halide generated vector code was largely unaffected by overlapped vs separated.

This effect is probably highly dependent on the CPU Cache architecture. Maybe your test-CPU can handle cache invalidation better. Or by any chance did you benchmark independent of SDL with separate input and output buffers?

I tested on an i7-7700K
